mlflow/sklearn/__init__.py

```diff
@@ -455,6 +460,323 @@ def load_model(model_uri):
+
+            def __exit__(inner_self, exc_type, exc_val, exc_tb):
+                self._log_post_training_metrics_enabled = inner_self.old_status
+
+        return LogPostTrainingMetricsDisabledScope()
+
+    @staticmethod
+    def get_run_id_for_model(model):
+        return getattr(model, '_mlflow_run_id', None)
+
+    @staticmethod
+    def is_metrics_value_loggable(metric_value):
+        """
+        check whether the specified `metric_value` is a numeric value which can be logged
+        as an MLflow metric.
+        """
+        return isinstance(metric_value, (int, float, np.number)) and \
+            not isinstance(metric_value, (bool, np.bool))
+
+    def register_model(self, model, run_id):
+        """
+        In `patched_fit`, we need register the model with the run_id used in `patched_fit`
+        So that in following metric autologging, the metric will be logged into the registered
+        run_id
+        """
+        model._mlflow_run_id = run_id
+
+    @staticmethod
+    def gen_name_with_index(name, index):
+        assert index >= 0
+        if index == 0:
+            return name
+        else:
+            # Use '-' as the separator between name and index,
+            # The '-' is not valid character in python var name
+            # so it can prevent name conflicts after appending index.
+            return f'{name}-{index + 1}'
+
+    def register_eval_dataset(self, model, eval_dataset):
+        """
+        Register eval dataset into eval_dataset_info_map, it will do:
+         1. inspect eval dataset var name.
+         2. check whether eval_dataset_info_map already registered this eval dataset.
+            will check by object id and comparing row samples.
+            Note: only check object id is not sufficient, id may be reused.
+         3. register eval dataset (with id and row samples information)
+         4. return eval dataset name with index.
+
+        Note: this method include inspecting argument variable name.
+         So should be called directly from the "patched method", to ensure it capture
+         correct argument variable name.
+        """
+        eval_dataset_name = _inspect_original_var_name(
+            eval_dataset, fallback_name='unknown_dataset')
+        eval_dataset_id = id(eval_dataset)
+
+        run_id = self.get_run_id_for_model(model)
+        registered_dataset_list = self._eval_dataset_info_map[run_id][eval_dataset_name]
+
+        index = len(registered_dataset_list)
+        for i, id_i in enumerate(registered_dataset_list):
+            if eval_dataset_id == id_i:
+                index = i
+                break
+
+        if index == len(registered_dataset_list):
+            # register new eval dataset
+            registered_dataset_list.append(eval_dataset_id)
+
+        return self.gen_name_with_index(eval_dataset_name, index)
+
+    def register_prediction_result(self, model, eval_dataset_name, predict_result):
+        """
```

Comment:
Can we remove the `model` parameter and have the caller pass in the `run_id` instead? In `patched_predict` we'll fetch the run ID from the model and pass it to this function:

```
         run_id = self.get_run_id_for_model(model)
         status.register_prediction_result(
                run_id, eval_dataset_name, predict_result)
```

This makes it clearer than the prediction result is being associated with a dataset name and a run ID, rather than a dataset name and a model.